File: translate_to_other_programming_language/bardapi.swift
import Foundation

class Bard {
    static let HEADERS: [String: String] = [
        "Host": "bard.google.com",
        "X-Same-Domain": "1",
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36",
        "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
        "Origin": "https://bard.google.com",
        "Referer": "https://bard.google.com/"
    ]
    
    let token: String?
    let timeout: TimeInterval
    let proxies: [String: String]?
    let session: URLSession
    
    private var reqid: Int
    private var conversationId: String
    private var responseId: String
    private var choiceId: String
    private var snim0e: String
    
    init(token: String? = nil, timeout: TimeInterval = 20, proxies: [String: String]? = nil, session: URLSession? = nil) {
        self.token = token ?? ProcessInfo.processInfo.environment["_BARD_API_KEY"]
        self.timeout = timeout
        self.proxies = proxies
        self.session = session ?? URLSession.shared
        
        self.reqid = Int.random(in: 0..<10_000)
        self.conversationId = ""
        self.responseId = ""
        self.choiceId = ""
        self.snim0e = ""
    }
    
    private func getSNlM0e(completion: @escaping (Result<String, Error>) -> Void) {
        guard let token = token, token.hasSuffix(".") else {
            completion(.failure(BardError.invalidToken))
            return
        }
        
        guard let url = URL(string: "https://bard.google.com/") else {
            completion(.failure(BardError.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.timeoutInterval = timeout
        request.allHTTPHeaderFields = Bard.HEADERS
        request.httpMethod = "GET"
        request.setValue(token, forHTTPHeaderField: "__Secure-1PSID")
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(BardError.invalidResponse))
                return
            }
            
            guard let data = data,
                  let htmlString = String(data: data, encoding: .utf8),
                  let snim0e = Bard.extractSNlM0e(from: htmlString) else {
                completion(.failure(BardError.snim0eNotFound))
                return
            }
            
            completion(.success(snim0e))
        }
        
        task.resume()
    }
    
    private static func extractSNlM0e(from htmlString: String) -> String? {
        let pattern = #"SNlM0e":"(.*?)""#
        let regex = try! NSRegularExpression(pattern: pattern)
        if let match = regex.firstMatch(in: htmlString, range: NSRange(htmlString.startIndex..., in: htmlString)),
           let range = Range(match.range(at: 1), in: htmlString) {
            return String(htmlString[range])
        }
        return nil
    }
    
    private func generateRequest(inputText: String, completion: @escaping (Result<[String: Any], Error>) -> Void) {
        let params = [
            "bl": "boq_assistant-bard-web-server_20230419.00_p1",
            "_reqid": "\(reqid)",
            "rt": "c"
        ]
        
        let inputTextStruct = [
            [inputText],
            NSNull(),
            [conversationId, responseId, choiceId]
        ]
        
        let data = [
            "f.req": "[null, \(Bard.jsonString(inputTextStruct))]",
            "at": snim0e
        ]
        
        guard let url = URL(string: "https://bard.google.com/_/BardChatUi/data/assistant.lamda.BardFrontendService/StreamGenerate") else {
            completion(.failure(BardError.invalidURL))
            return
        }
        
        var request = URLRequest(url: url)
        request.timeoutInterval = timeout
        request.allHTTPHeaderFields = Bard.HEADERS
        request.httpMethod = "POST"
        request.httpBody = try! JSONSerialization.data(withJSONObject: data)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.url = URL(string: url.absoluteString + "?" + params.map { "\($0.key)=\($0.value)" }.joined(separator: "&"))
        
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(BardError.invalidResponse))
                return
            }
            
            guard let data = data,
                  let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []),
                  let responseArray = jsonResponse as? [Any],
                  let respDict = responseArray[3] as? [String: Any] else {
                completion(.failure(BardError.invalidResponseData))
                return
            }
            
            completion(.success(respDict))
        }
        
        task.resume()
    }
    
    func getAnswer(inputText: String, completion: @escaping (Result<[String: Any], Error>) -> Void) {
        getSNlM0e { [weak self] result in
            switch result {
            case .success(let snim0e):
                self?.snim0e = snim0e
                self?.generateRequest(inputText: inputText, completion: completion)
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    private static func jsonString(_ object: Any) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: object) else {
            return ""
        }
        return String(data: data, encoding: .utf8) ?? ""
    }
}

enum BardError: Error {
    case invalidToken
    case invalidURL
    case invalidResponse
    case snim0eNotFound
    case invalidResponseData
}

File: translate_to_other_programming_language/bardapi.js
const os = require('os');
const randomstring = require('randomstring');
const axios = require('axios');

class Bard {
  HEADERS = {
    Host: 'bard.google.com',
    'X-Same-Domain': '1',
    'User-Agent':
      'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36',
    'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
    Origin: 'https://bard.google.com',
    Referer: 'https://bard.google.com/',
  };

  constructor(token = null, timeout = 20, proxies = null, session = null) {
    this.token = token || process.env._BARD_API_KEY;
    this.proxies = proxies;
    this.timeout = timeout;
    this._reqid = parseInt(randomstring.generate({ length: 4, charset: 'numeric' }));
    this.conversation_id = '';
    this.response_id = '';
    this.choice_id = '';
    this.session = session || axios.create({
      headers: this.HEADERS,
      withCredentials: true,
      timeout: this.timeout * 1000,
      proxy: this.proxies,
    });
    this.SNlM0e = null;
  }

  async _get_snim0e() {
    if (!this.token || this.token[this.token.length - 1] !== '.') {
      throw new Error('__Secure-1PSID value must end with a single dot. Enter correct __Secure-1PSID value.');
    }
    const resp = await this.session.get('https://bard.google.com/');
    if (resp.status !== 200) {
      throw new Error(`Response code not 200. Response Status is ${resp.status}`);
    }
    const snim0eMatch = resp.data.match(/SNlM0e":"(.*?)"/);
    if (!snim0eMatch) {
      throw new Error('SNlM0e value not found in response. Check __Secure-1PSID value.');
    }
    this.SNlM0e = snim0eMatch[1];
  }

  async get_answer(input_text) {
    if (!this.SNlM0e) {
      await this._get_snim0e();
    }

    const params = {
      bl: 'boq_assistant-bard-web-server_20230419.00_p1',
      _reqid: this._reqid.toString(),
      rt: 'c',
    };
    const input_text_struct = [
      [input_text],
      null,
      [this.conversation_id, this.response_id, this.choice_id],
    ];
    const data = {
      'f.req': JSON.stringify([null, JSON.stringify(input_text_struct)]),
      at: this.SNlM0e,
    };
    const resp = await this.session.post(
      'https://bard.google.com/_/BardChatUi/data/assistant.lamda.BardFrontendService/StreamGenerate',
      { params, data }
    );
    const [, , resp_dict] = JSON.parse(resp.data.split('\n')[3]);

    if (!resp_dict) {
      return { content: `Response Error: ${resp.data}` };
    }
    const parsed_answer = JSON.parse(resp_dict);
    const bard_answer = {
      content: parsed_answer[0][0],
      conversation_id: parsed_answer[1][0],
      response_id: parsed_answer[1][1],
      factualityQueries: parsed_answer[3],
      textQuery: parsed_answer[2][0] || '',
      choices: parsed_answer[4].map(([id, content]) => ({ id, content })),
    };
    this.conversation_id = bard_answer.conversation_id;
    this.response_id = bard_answer.response_id;
    this.choice_id = bard_answer.choices[0].id;
    this._reqid += 100000;

    return bard_answer;
  }
}

module.exports = Bard;
